security/pkg/nodeagent/caclient/providers/citadel/client.go

```diff
@@ -112,58 +117,37 @@ func (c *CitadelClient) CSRSign(csrPEM []byte, certValidTTLInSec int64) ([]strin
 }
 
 func (c *CitadelClient) getTLSDialOption() (grpc.DialOption, error) {
-	// Load the TLS root certificate from the specified file.
-	// Create a certificate pool
-	var certPool *x509.CertPool
-	var err error
-	if c.caTLSRootCert == nil {
-		// No explicit certificate - assume the citadel-compatible server uses a public cert
-		certPool, err = x509.SystemCertPool()
-		if err != nil {
-			return nil, err
-		}
-		citadelClientLog.Info("Citadel client using public DNS: ", c.opts.CAEndpoint)
-	} else {
-		certPool = x509.NewCertPool()
-		ok := certPool.AppendCertsFromPEM(c.caTLSRootCert)
-		if !ok {
-			return nil, fmt.Errorf("failed to append certificates")
-		}
-		citadelClientLog.Info("Citadel client using custom root cert: ", c.opts.CAEndpoint)
+	certPool, err := getRootCertificate(c.tlsOpts.RootCert)
+	if err != nil {
+		return nil, err
 	}
 	var certificate tls.Certificate
 	config := tls.Config{
-		Certificates: []tls.Certificate{certificate},
 		GetClientCertificate: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
-			if c.opts.ProvCert != "" {
-				// Load the certificate from disk
-				certificate, err = tls.LoadX509KeyPair(
-					filepath.Join(c.opts.ProvCert, "cert-chain.pem"),
-					filepath.Join(c.opts.ProvCert, "key.pem"))
-
-				if err != nil {
-					// we will return an empty cert so that when user sets the Prov cert path
-					// but not have such cert in the file path we use the token to provide verification
-					// instead of just broken the workflow
-					citadelClientLog.Warnf("cannot load key pair, using token instead: %v", err)
-					return &certificate, nil
-				}
+			key, cert := c.tlsOpts.Key, c.tlsOpts.Cert
+			if cert != "" {
 				var isExpired bool
-				isExpired, err = c.isCertExpired(filepath.Join(c.opts.ProvCert, "cert-chain.pem"))
+				isExpired, err = c.isCertExpired(cert)
 				if err != nil {
 					citadelClientLog.Warnf("cannot parse the cert chain, using token instead: %v", err)
-					return &tls.Certificate{}, nil
+					return &certificate, nil
```

Comment:
This seems to change behavior? Before it always returns empty cert. Now it may return `certificate` which is set by previous call to this function

Reply:
Same on L137

Reply:
sorry, my mistake, I meant to define `certificate` under the callback